packages/react-server/core/TritonAgent/CacheEntry.js

```diff
@@ -0,0 +1,317 @@
+			// call setResponse to resolve the deferred
+			if (this.res) {
+				this.setResponse(this.res);
+			} else if (this.err) {
+				this.setError(this.err);
+			}
+			logger.debug(`Rehydrating resolved url to cache: ${this.url}`);
+		} else {
+			logger.debug(`Rehydrating pending url to cache without data: ${this.url}`);
+		}
+	}
+
+	_rehydrateResponse (res) {
+		if (!res) return res;
+
+		if (res._hasBody) {
+			// re-parse the text of the response body serialized by the server.
+			// if the body wasn't in a known format, it will have been included directly
+
+			var parse = responseBodyParsers[res.type];
+			if (!parse) {
+				logger.warning(`Unparseable content type for ${this.url}: ${res.type}, but response._hasBody was true. (This may be a bug in TritonAgent)`);
+			}
+			res.body = parse && res.text && res.text.length
+				? parse(res.text)
+				: null;
+			delete res._hasBody;
+		}
+
+		return res;
+	}
+
+	setResponse (res) {
+		// TODO: store superagent response? or body? or payload?
+
+		if (SERVER_SIDE){
+
+			// Pull out the pieces of the response we care about.
+			// This would be a NOOP client-side, so we'll skip it.
+			res = this._trimResponseData(res);
+		}
+
+		// Stash away a reference to the response.
+		this.res    = res;
+		this.loaded = true;
+
+		if (SERVER_SIDE){
+
+			// Deep copy.
+			//
+			// Leave ourselves with a clean copy of the original
+			// response regardless of what mutation might happen
+			// once stores get ahold of it.
+			//
+			// This is important to ensure that we provide the same
+			// data from the cache when we wake up in the browser
+			// as we initially provide on the server.
+			//
+			res = JSON.parse(JSON.stringify(res));
+		}
+
+		this.dfd.resolve(res);
+	}
+
+	setError (err) {
+
+		if (SERVER_SIDE) {
+
+			// If the error was caused by a server response, trim it
+			// and serialize it like a regular response
+			if (err && err.response) {
+				err.response = this._trimResponseData(err.response);
+			}
+		}
+
+		this.err = err;
+		this.loaded = true;
+
+		if (SERVER_SIDE) {
+			// Deep copy, to make sure nobody plays with the
+			// object we put in the cache
+			err = JSON.parse(JSON.stringify(err));
+		}
+
+		this.dfd.reject(err);
+	}
+
+	whenDataReady () {
+		if (SERVER_SIDE) {
+			// server-side, we increment the number of requesters
+			// we expect to retrieve the data on the frontend
+			this.requesters += 1;
+			return this.dfd.promise;
+		} else {
+			// client-side, whenever someone retrieves data from the cache,
+			// we decrement the number of retrievals expected, and when we
+			// hit zero, remove the cache entry.
+			return this._requesterDecrementingPromise(this.dfd.promise);
+		}
+	}
+
+	// for internal (triton middleware) calls
+	whenDataReadyInternal () {
+		return this.dfd.promise;
+	}
+
+	decrementRequesters () {
+		logger.debug("Decrementing: " + this.url);
+		this.requesters -= 1;
+
+		if (this.requesters === 0) {
+			this.cache._removeEntry(this);
```

Comment:
for the uncached responses, `this.cache` could be null here on the client-side. (Not that I'd expect anybody to want to do that, since things on the client-side aren't cached anyway...)